src/transform-sdk/js/main.cc

```diff
@@ -315,8 +315,399 @@ std::expected<qjs::value, qjs::exception> make_write_event(
+      "lookupLatestSchema");
+    builder.method<&schema_registry_client::create_schema>("createSchema");
+    return builder.build();
+}
+
+redpanda::bytes_view value_as_bytes_view(const qjs::value& val) {
+    if (val.is_string()) {
+        auto data = val.string_data();
+        return redpanda::bytes_view{data.view()};
+    }
+    if (val.is_array_buffer()) {
+        auto data = val.array_buffer_data();
+        return redpanda::bytes_view{data.data(), data.size()};
+    }
+    if (val.is_uint8_array()) {
+        auto data = val.uint8_array_data();
+        return redpanda::bytes_view{data.data(), data.size()};
+    }
+    return redpanda::bytes_view{};
+}
+
+std::expected<qjs::value, qjs::exception>
+decode_schema_id_impl(JSContext* ctx, qjs::value& arg) {
+    auto data = value_as_bytes_view(arg);
+    auto obj = qjs::value::object(ctx);
+    return redpanda::sr::decode_schema_id(data)
+      .transform_error([ctx](std::error_code err) {
+          return qjs::exception::make(
+            ctx, std::format("Failed to decode schema ID: {}", err.message()));
+      })
+      .and_then(
+        [ctx,
+         &obj](std::pair<redpanda::sr::schema_id, redpanda::bytes_view> sid) {
+            return obj.set_property("id", qjs::value::integer(ctx, sid.first));
+        })
+      .and_then(
+        [ctx,
+         &arg](std::monostate) -> std::expected<qjs::value, qjs::exception> {
+            // NOTE(oren): returning a view back into the target buffer might
+            // be a terrible idea. Needs a good think.
+            constexpr size_t HEADER_SIZE = 5;
+            if (arg.is_string()) {
+                return qjs::value::string(
+                  ctx, arg.string_data().view().substr(HEADER_SIZE));
+            }
+            if (arg.is_array_buffer()) {
+                return qjs::value::array_buffer(
+                  ctx, arg.array_buffer_data().subspan(HEADER_SIZE));
+            }
+            if (arg.is_uint8_array()) {
+                return qjs::value::uint8_array(
+                  ctx, arg.uint8_array_data().subspan(HEADER_SIZE));
+            }
+            // NOTE(oren): we should never reach here (covered by earlier
+            // checks)
+            return std::unexpected(qjs::exception::make(
+              ctx, "Unexpected type for schema ID decode"));
+        })
+      .and_then([&obj](const qjs::value& rest) {
+          return obj.set_property("rest", rest);
+      })
+      .transform([&obj](std::monostate) { return std::move(obj); });
+}
+
+std::expected<qjs::value, qjs::exception> encode_schema_id_impl(
+  JSContext* ctx,
+  const qjs::value& sid, // NOLINT(*-swappable-parameters)
+  const qjs::value& buf  // NOLINT(*-swappable-parameters)
+) {
+    auto orig = value_as_bytes_view(buf);
+    redpanda::bytes encoded = redpanda::sr::encode_schema_id(
+      redpanda::sr::schema_id{sid.as_integer()}, orig);
+    return qjs::value::uint8_array_copy(ctx, encoded);
```

Comment:
>can't we pass the original bytes

not sure how this would work mechanically. we get a vector back from `encode_schema_id`, so I'm fairly certain you have to copy those out into an unmanaged buffer anyway.

could return a `unique_ptr<uint8_t*>` from `encode_schema_id` or something else that allows releasing the underlying memory. am I missing something? maybe overload `encode_schema_id` to take a view into memory allocated at the call site?

generally agree that we can punt on this though. will add the TODO.

Reply:
You would have to "move" the vector to a heap allocated pointer (no copies of the data) and make that the opaque value in the JSRuntime that is deleted when it's done. No need to do now.

Reply:
yeah, i tried that... though you know what, i may have been going through the wrong pointer param 🤦

Reply:
probably fast follow if I have time